addons/udes_stock/models/stock_move_line.py

```diff
@@ -252,3 +252,139 @@ def get_move_lines_ordered_by(self, domain, aux_domain=None, order="id"):
+        incoming = mls.u_picking_type_id.code == "incoming"
+        res = defaultdict(dict)
+        if not product_ids:
+            if package:
+                pack_mls = package._get_current_move_lines()
+                # TODO: this is probably where swap package should go
+                if (pack_mls & mls) != pack_mls:
+                    raise ValidationError(
+                        _("All package %s move lines cannot be found in picking %s") %
+                        (package.name, mls.picking_id.name)
+                    )
+                res[pack_mls] = vals
+            else:
+                res[mls] = vals
+            # Set to empty list to bypass the for
+            product_ids = []
+        for prod in product_ids:
+            product_barcode = prod["barcode"]
+            product = Product.get_or_create(product_barcode)
+            quantity = prod["uom_qty"]
+            lot_names = prod.get("lot_names", [])
+            product_tracking = product.tracking
+            is_serial = product_tracking == "serial"
+            is_lot = product_tracking == "lot"
+            if is_serial and len(lot_names) != quantity:
+                raise ValidationError(
+                    _("Not enough serial numbers provided.")
+                )
+            elif is_lot and len(lot_names) != 1:
+                raise ValidationError(
+                    _("Too many lots provided.")
+                )
+            qty_done = quantity
+            if product_tracking != "none":
+                quantity_fulfilled = 0
+                if is_serial:
+                    qty_done = 1
+                for lot_name in lot_names:
+                    lot_name_val = lot_name
+                    # If it is incoming we don't need to match the lot
+                    if incoming:
+                        lot_name = None
+                    else:
+                        lot_name_val = None
+                    prod_mls, new_ml = mls._find_move_lines(
+                        qty_done, product, package, lot_name, location
+                    )
+                    prod_dict = {
+                        "qty_done": qty_done,
+                        "lot_name": lot_name_val
+                    }
+                    prod_dict.update(vals)
+                    res[prod_mls] = prod_dict
+                    quantity_fulfilled += qty_done
+                    mls -= prod_mls
+                    if new_ml:
+                        mls |= new_ml
+                    if quantity_fulfilled == quantity:
+                        break
+            else:
+                prod_mls, new_ml = mls._find_move_lines(
+                    qty_done, product, package, None, location
+                )
+                prod_dict = {"qty_done": qty_done}
+                prod_dict.update(vals)
+                res[prod_mls] = prod_dict
+                mls -= prod_mls
+                if new_ml:
+                    mls |= new_ml
+        return res
```

Comment:
I've recently seen it suggested that once a `defaultdict` has been populated, it should be converted to a standard `dict` so that it isn't possible to accidentally add more keys.  So we could return `dict(res)` here, assuming the caller isn't depending on the `defaultdict` features.  If you think that kind of defensive programming is worth it.